docs/ref/models/constraints.txt

```diff
@@ -95,3 +95,13 @@ ensures that each user only has one draft.
 
 These conditions have the same database restrictions as
 :attr:`Index.condition`.
+
+.. admonition:: Model validation with Meta.constraints
+
+    In general constraints are ``not`` checked during ``full_clean()``, and so do
+    not raise ``ValidationError``\s. Rather you'll get a database integrity
+    error on ``save()``. ``UniqueConstraint``\s are different in this regard,
+    in that they leverage the existing ``validate_unique()`` logic, and thus
+    ``ValidationError`` is raised during model validation when the
+    ``UniqueConstraint`` is violated.
```

Comment:
I don't think that such detailed explanation is necessary, moreover you added admonition about all constraints types under `UniqueConstraint` description. I would add a short sentence to both class, e.g. 

- to the `UniqueConstraint`:
```
The `ValidationError` is raised during model validation when the
``UniqueConstraint`` is violated.
```
- to the `CheckConstraint`:
```
The database `IntegrityError` is raised during ``model.save()`` when the
``CheckConstraint`` is violated.
```

Reply:
OK, I found Carlton's [comment](https://github.com/django/django/pull/11475#discussion_r293745091). You can ignore my comment for now, let's leave it to him.

Reply:
Sure felixxm.

Reply:
Right, so I'd put this up the top in it's own section (or admonition) _Validation of Constraints_ (or similar).

* The point being that it's **unexpected** that different kinds of errors are raised depending on the constraint type. 
* I think it's because there's a bit of missing API: constraints need a hook for an equivalent bit of Python, which, if implemented, could be called in `full_clean()` or form validation maybe to raise a `ValidationError`, just as the `UniqueConstraint` already does. 

It's not quite right to say: 

> The `ValidationError` is raised during model validation when the
``UniqueConstraint`` is violated.

... because if you bypass `full_clean()` you'll get the `IntegrityError` from the DB. So we still need to explain about the two-stage validation there. 

As it is now, whatever text we add will be a bit _naff_, because you **can't** yet validate a constraint in Python, but if we can add such a hook, it should eventually read, _"Constraints allow a two-part validation... and here's how you enable it..."_

//cc @Ian-Foote

Reply:
Fixed location of admonition and added info about two-stage validation of UniqueConstraint.